internal/auth/password.go

```diff
@@ -0,0 +1,116 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package auth
+
+import (
+	"crypto/rand"
+	"crypto/sha512"
+	"encoding/base64"
+	"fmt"
+	"io"
+
+	"github.com/juju/errors"
+	"golang.org/x/crypto/pbkdf2"
+)
+
+// Password hides and protects a plain text passwords in Juju from accidentally
+// being consumed or printed to a log.
+type Password struct {
+	// password is the plain text password being protected
+	password []byte
+}
+
+const (
+	// ErrorPasswordNotValid is used when a password has failed validation.
+	ErrorPasswordNotValid = errors.ConstError("password not valid")
+
+	// defaultSaltLength is the default number of bytes that we randomly
+	// generate for new salts. The pbldf2 guidelines is to use 8 bytes for salt.
+	defaultSaltLength = 12
+
+	// pbkdf2Iterations is the number iterations to perform for pbkdf2. The
+	// more iterations performed the harder brute forcing is. Each iteration
+	// repeatedly hashes the password adding computational complexity.
+	//
+	// Warning updating this value will break already established passwords on
+	// the old value!
+	pbkdf2Iterations = 8192
+
+	// pbkdf2KeySize is the size of the produced key by pbkdf2KeySize, expressed
+	// in number of bytes.
+	//
+	// Warning updating this value will break already established passwords on
+	// the old value!
+	pbkdf2KeySize = 18
+
+	// maxPasswordSizeBytes is the maximum size of a password that we will
+	// tolerate.
+	maxPasswordSizeBytes = 1024 // 1KB is the max we are going to allow.
+)
+
+// HasPassword takes a password and corresponding salt to produce a hash of the
+// password. The resultant hash is safe for persistence and comparison.
+// If the salt provided to password hash is empty then a error satisfying
+// errors.NotValid is returned. If the password does not pass validation a error
+// satisfying ErrorPasswordNotValid will be returned.
+func HashPassword(p Password, salt []byte) (string, error) {
+	if err := p.Validate(); err != nil {
+		return "", fmt.Errorf("hashing password: %w", err)
+	}
+	if len(salt) == 0 {
+		return "", fmt.Errorf(
+			"%w password salt cannot be empty for password hashing", errors.NotValid,
+		)
+	}
+	h := pbkdf2.Key(p.password, salt, pbkdf2Iterations, pbkdf2KeySize, sha512.New)
+	return base64.StdEncoding.EncodeToString(h), nil
+}
+
+// NewSalt generates a new random password salt for use with password hashing.
+func NewSalt() ([]byte, error) {
+	buf := [defaultSaltLength]byte{}
+	_, err := io.ReadFull(rand.Reader, buf[:])
+	if err != nil {
+		return nil, fmt.Errorf("generating random bytes for new password salt: %w", err)
+	}
+
+	dst := make([]byte, base64.StdEncoding.EncodedLen(len(buf)))
+	base64.StdEncoding.Encode(dst, buf[:])
+	return dst, nil
+}
+
+// NewPassword returns a Password struct wrapping the plain text password.
+func NewPassword(p string) Password {
```

Comment:
I think we should use a []rune or []byte here, to ensure the password can be zero'd after the hashing is completed.

Reply:
It is still going to be pass by ref and would require a big change in the apiserver. Happy to look at this but it's going to require a lot of magic on the APIServer end that we just can't do in this PR.